lib/request.ex

```diff
@@ -14,17 +14,37 @@ defmodule HTTPServer.Request do
   def parse_request(message) do
     [request_data | body] = message |> String.split("#{@carriage_return}#{@carriage_return}")
     [first | headers] = request_data |> String.split("#{@carriage_return}")
+    headers = parse_headers(headers, %{})
+    body = parse_body(hd(body), headers)
     [method, path, resource] = first |> String.split(" ")
 
     %__MODULE__{
       method: method,
       path: path,
       resource: resource,
-      headers: parse_headers(headers, %{}),
-      body: hd(body)
+      headers: headers,
+      body: body
     }
   end
 
+  defp parse_body(body, headers) do
+    if is_content_type_not_nil(headers) && is_content_type_json(headers) do
+      {:ok, parsed_body} = JSON.decode(body)
+      parsed_body
+    else
+      body
+    end
+  end
+
+  defp is_content_type_not_nil(headers) do
+    headers["Content-Type"] != nil
+  end
+
+  defp is_content_type_json(headers) do
+    [type | _parameter] = headers["Content-Type"] |> String.split(";")
+    type == "application/json"
+  end
```

Comment:
I looked at this and immediately thought "if the content type is `application/json` then it is obviously not `nil`, and therefore you do not need the first check. Is that true? and can this be simplified?

Reply:
`is_content_type_json/1` will throw an error at `|> String.split(";")` if there is no content-type header, so it was sort of my solution to ensure that anything being passed to that function is definitely not null. I also thought it helped with clarity instead of handling all of that within `is_content_type_json/1`.

Reply:
I would make the argument that it does exist in `is_content_type_json/1` because that function would error without it.  Including it in `is_content_type_json/1` would not make it do more than one thing, it would just ensure that it does that one thing well.

Reply:
at least the call to it, maybe having the logic separated out like you do makes sense.

Reply:
```suggestion
  defp parse_body(body, headers) do
    if is_content_type_json(headers) do
      {:ok, parsed_body} = JSON.decode(body)
      parsed_body
    else
      body
    end
  end
  
  defp is_content_type_json(headers) do
    unless headers["Content-Type"] == nil do
      [type | _parameter] = headers["Content-Type"] |> String.split(";")
      type == "application/json"
    else
      false
    end
  end
```

Another option combining those two methods into one unless statement

Reply:
Would this work and remove the need for the `nil` check?

```elixir
defp is_content_type_json(headers) do
  case headers["Content-Type"] |> String.split(";")
    [type | _paramenter] ->
      type == "application/json"

    _ ->
      False
  end
end
```

Reply:
> Would this work and remove the need for the nil check?

No this one's still throwing this error:
```
14:00:58.116 [error] Process #PID<0.317.0> raised an exception
** (FunctionClauseError) no function clause matching in String.split/3
    (elixir 1.14.5) lib/string.ex:479: String.split(nil, ";", [])
    (http_server 0.1.0) lib/request.ex:40: HTTPServer.Request.is_content_type_json/1
    (http_server 0.1.0) lib/request.ex:31: HTTPServer.Request.parse_body/2
    (http_server 0.1.0) lib/request.ex:18: HTTPServer.Request.parse_request/1
    (http_server 0.1.0) lib/http_server.ex:27: HTTPServer.serve/1
```

I could extract it out though to `headers["Content-Type"] != nil` as you mentioned before!